arch/power/compare256_power9.c

```diff
@@ -0,0 +1,58 @@
+/* compare256_power9.c - Power9 version of compare256
+ * Copyright (C) 2019 Matheus Castanho <msc@linux.ibm.com>, IBM
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#ifdef POWER9
+#include <altivec.h>
+#include "../../zbuild.h"
+#include "../../zendian.h"
+
+static inline uint32_t compare256_power9_static(const uint8_t *src0, const uint8_t *src1) {
+    vector unsigned char vsrc0, vsrc1, vc;
+    uint32_t len = 0, cmplen;
+
+    do {
+        vsrc0  = *((vector unsigned char *)src0);
+        vsrc1 = *((vector unsigned char *)src1);
+
+        /* Compare 16 bytes at a time. Each byte of vc will be either
+        * all ones or all zeroes, depending on the result of the comparison. */
+        vc = (vector unsigned char)vec_cmpne(vsrc0, vsrc1);
+
+        /* Since the index of matching bytes will contain only zeroes
+         * on vc (since we used cmpne), counting the number of consecutive
+         * bytes where LSB == 0 is the same as counting the length of the match. */
+#if BYTE_ORDER == LITTLE_ENDIAN
+        cmplen = __builtin_vec_vctzlsbb(vc);
+        //asm volatile("vctzlsbb %0, %1\n\t" : "=r" (len) : "v" (vc));
```

Comment:
@mscastanho do you know why the assembly is the same in both branches? I saw this on your zlib PR too. https://github.com/madler/zlib/pull/459/files#r852564423

Reply:
Good lord, can somebody from IBM tell them they need to publish more documentation on their ISA?  It's gotten to the point that Google has no idea what I'm talking about when I a search for the exact instruction name, let alone the intrinsic.  From what I can tell, when I finally found what the acronym was in a patent document of all things, it counts the number of bytes in the vector for which the least significant bit is zero, trailing or leading depending on whatever direction you pick.  Which is basically an instruction that shortcuts the whole "movemask + tzcnt" sequence in x86 land (probably because up until more modern versions of POWER, moving between GPRs and vectors wasn't an option).

I'd like to know what the minimum version of POWER is for this instruction to be supported since I'd like to implement this on altivec hardware.  A GCC include _kind of_ indicates that it might just work, but possibly only on PPC64?  I do have a 64 bit OS and a 64 bit PPC CPU, but a good number of PPC systems of this era are 32 bit only.  A compile time #ifdef I guess would take care of that...

Reply:
It looks like where it is in this includes indicates that it may not be supported until POWER9:
https://gist.github.com/Cyan4973/1f87f02b2c96a37470205baaf12e1407

And only if compiling with PPC64 support, as well.  So, the < POWER9 variants will need to use a sequence that takes a trip through stack memory, barring some other clever mechanism to find the first set bit in a vector.

Reply:
We really don't need to worry about trying to support every iteration of POWER architecture with vector optimized variants... Even POWER9 doesn't have all the useful features as far as I know, as POWER10 added some useful instructions.

Until someone or some company donates us hardware that we can use to test and benchmark every possible architecture and CPU variant, we can only merge in pull requests that we actually can verify works as intended.

Reply:
Right, but arguably POWER4 and it's derivatives from Apple are going to be a bit more commonplace due to their ubiquity and wide scale marketing in the consumer market.  Though, workstation hardware from IBM and Talos may eventually exceed that stockpile.  It'd sure help if they didn't cost as much as cars.  Talos is a step in the right direction, but it's still a bit steep.

Reply:
@nmoinvaz  actually, strictly speaking I don't think it actually matters if comparing the lead or the trail given that for equality either bit would be zero of the byte, hah.

Reply:
Well...perhaps not.  It should either be one way or the other but the fact that this code switches based on endianness tells me that POWER implicitly swaps bytes somehow when loading a vector.  That more or less jives with their unaligned load semantics from what I recall a while back when reading about some of the newer POWER variants.  This is, of course, kind of confusing, you would think that loading a vector of characters would not have a implicit endianness switch with it.

Reply:
Even on little-endian architectures, the vector processing unit is usually big-endian, so element order is reversed. This is separate from byte order inside larger elements.

Some architectures can load and reverse the element order (so it matches memory order) using single instruction.

Reply:
@KungFuJesus The go-to place for the Power ISA docs is the OpenPower Foundation website: https://openpower.foundation/specifications/isa/

And you're right, `vctzlsbb`/`vclzlsbb` were added in ISA version 3.0 (easier to verify in the tables on Appendices E-H of the ISA), which was first implemented by Power 9, so they won't be available for earlier processors.

> POWER implicitly swaps bytes somehow when loading a vector.

Yep. The way the loaded bytes are placed in the vector changes if it's BE or LE, that's why the code counts either trailing zeroes or leading zeroes, depending on the endianness.

Reply:
I'll keep that in mind for the future.  It's a bit strange to me that google isn't indexing content from there very well - you'd think that'd be the most relevant search result.  

On a sadder note it seems that the informatimago mirror that had a gold mine of the old Apple optimization guides for the PPC970 and the 7400/7450 is now gone.  Archive.org seems to have captured a good amount of it but I wish I had stashed that before it disappeared.